server/user.js

```diff
@@ -0,0 +1,330 @@
+/**
+ * @fileoverview user function handler
+ */
+var fs = require('fs');
+var CryptoJS = require('crypto-js');
+
+var log = require('./log.js');
+var utils = require('./utils.js');
+var database = require('./database.js');
+
+/** @type {user} */
+module.exports = user;
+
+/**
+ * @constructor
+ */
+function user() {}
+
+/** @enum {string} */
+user.QueryType = {
+  SIGNUP: 'sign-up',
+  SIGNIN: 'sign-in',
+  AUTOSIGNIN: 'auto-sign-in'
+};
+
+/**
+ * Database name of user information.
+ * @const {string}
+ */
+user.userInfoDbName = 'userInfo';
+
+/**
+ * Database name of session.
+ * @const {string}
+ */
+user.sessionDbName = 'session';
+
+/**
+ * Expire time of cookie.
+ * @const {number}
+ */
+user.cookieExpireTime = 24 * 60 * 60 * 1000;
+
+/** @const {RegExp} */
+user.VALID_EMAIL_REGEX =
+  /^([a-zA-Z0-9_\.\-])+\@(([a-zA-Z0-9\-])+\.)+([a-zA-Z0-9]{2,4})+$/;
+
+/** @const {RegExp} */
+user.VALID_USERNAME_REGEX = /^\w{6,}$/;
+
+/** @const {RegExp} */
+user.VALID_PASSWORD_REGEX = /^\w{8,}$/;
+
+/**
+ * @typedef {{
+ *   email: (string|undefined),
+ *   username: string,
+ *   password: string,
+ *   confirmed: (boolean|undefined)
+ * }}
+ */
+user.Info;
+
+/**
+ * @typedef {{
+ *   username: (string|undefined),
+ *   sessionId: (string|undefined),
+ *   expiration: (number|undefined)
+ * }}
+ */
+user.Cookie;
+
+/**
+ * @typedef {{
+ *   error: string
+ * }}
+ */
+user.Error;
+
+/** @const */
+user.query = {};
+
+// Start public APIs
+/**
+ * @param {*|!user.Info} query
+ * @param {function((!user.Error|!user.Cookie))} callback Callback function.
+ */
+user.query.signIn = function(query, callback) {
+  var data;
+  var userInfo = {
+    username: query.username,
+    password: CryptoJS.SHA256(query.password).toString()
+  };
+  if (!user.validateUserInfo(userInfo.username, userInfo.password)) {
+    return;
```

Comment:
? I found that you were always misunderstanding the comments. I'm saying if the userInfo is invalidate, then it shall panic rather than return. Otherwise the server has no response.